File: OS/dfsm.py
class Process:
    def __init__(self,pid,burst_time,arrival_time):
        self.pid = pid 
        self.burst_time = burst_time
        self.arrival_time = arrival_time
        
def find_Waiting(processes):
    processes.sort(key= lambda x: x.arrival_time)
    curr_time = 0
    waiting_time = [0] * len(processes)
    
    for idx,process in enumerate(processes):
        if idx!=0:
            waiting_time[idx] = curr_time - process.arrival_time
        curr_time += process.burst_time
    
    print("List Of Waiting Time : ",waiting_time)
    print("Average waiting time is: ",sum(waiting_time)/len(processes))
    
    find_turnaround(processes,waiting_time)

def find_turnaround(processses,waiting_time):
    turn_around_time = [0] * len(processes)
    
    for idx,process in enumerate(processes):
        turn_around_time[idx] = process.burst_time + waiting_time[idx]
    
    print("List Of Turnaround Time : ",turn_around_time)
    print("Average Turnaround time is: ",sum(turn_around_time)/len(processes))
    

if __name__ == '__main__':
    processes = [Process(1,6,0),Process(2,8,1),Process(3,7,2),Process(4,3,3)]
    find_Waiting(processes)

File: OS/dfsm.c
#include <stdio.h>

struct Process
{
    int pid, arrival_time, burst_time;
};

int find_turnaround(struct Process *processes, int len, int *waiting_time);

int find(struct Process *processes, int len){
    int total = 0;
    for (int i = 0; i < len; i++)
    {
        total += processes[i].burst_time;
    }
    return total;
}

int find_turnaround(struct Process *processes, int len, int *waiting_time){
    int total = 0;
    printf("Turnaround Time For Each Process : ");
    for (int i = 0; i < len; i++)
    {
        printf("%d ", waiting_time[i] + processes[i].burst_time);
        total += waiting_time[i] + processes[i].burst_time;
    }
    printf("Average Turnaround Time : %d", total / len);
}

int find_avg_waiting(struct Process *processes, int len)
{
    int curr_waiting = 0;
    int waiting_time[len];

    waiting_time[0] = 0;

    for (int i = 1; i < len; i++)
    {
        curr_waiting = find(processes, i);
        waiting_time[i] = curr_waiting - processes[i].arrival_time;
    }

    int total = 0;
    printf("Average Waiting Time For Each Process : ");
    for(int i = 0; i < len; i++)
    {
        printf("%d ", waiting_time[i]);
        total+=waiting_time[i];
    }
    printf("Average waiting time : %d" , total/len);

    find_turnaround(processes,len,waiting_time);
}

int main()
{
    struct Process p1, p2, p3;
    p1 = (struct Process){1, 0, 3};
    p2 = (struct Process){2, 2, 6};
    p3 = (struct Process){3, 4, 5};

    struct Process processes[] = {p1, p2, p3};
    int len = sizeof(processes) / sizeof(processes[0]);
    find_avg_waiting(processes, len);

    return 0;
}